contracts/W3qProver.sol

```diff
@@ -0,0 +1,97 @@
+pragma solidity ^0.8.0;
+
+import "./interfaces/IW3qProver.sol";
+import "./lib/BlockDecoder.sol";
+import "./lib/MerklePatriciaProof.sol";
+import "./LightClient.sol";
+import "./interfaces/IW3qProver.sol";
+
+contract W3qProver is LightClient, IW3qProver {
+    using MerklePatriciaProof for bytes;
+    using BlockDecoder for bytes;
+    uint256 public override latestBlockHeight;
+
+    mapping(uint256 => bytes32) public override headHashes;
+    mapping(uint256 => BlockDecoder.HeadCore) public headCores;
+
+    constructor(uint256 _epochPeriod, address _staking) LightClient(_epochPeriod, _staking) {}
+
+    function initEpoch(
+        address[] memory _epochSigners,
+        uint256[] memory _epochVotingPowers,
+        uint256 height,
+        bytes32 headHash
+    ) public override(LightClient) onlyOwner {
+        _createEpochValidators(1, height, _epochSigners, _epochVotingPowers);
+        latestBlockHeight = height;
+        headHashes[height] = headHash;
+    }
+
+    function submitHead(
+        uint256 height,
+        bytes memory headBytes,
+        bytes memory commitBytes,
+        bool lookByIndex
+    ) public override(IW3qProver, LightClient) {
+        require(!blockExist(height), "block exist");
+
+        uint256 _epochIdx = getEpochIdx(height);
+        uint256 _position = _epochPosition(_epochIdx);
```

Comment:
May need to check whether `_position` is valid or not?  I understand that even without the check, the code may still work (as the validators should not sign an old block header).  But including the check makes sure that we have considered the case (or at least explicitly comment it here)